### PR TITLE
[#231] Modify local repository URL to the correct remote repository

### LINKED ISSUE #231: Clicking on ramps in ramp chart generated using local repos redirects to non-existent link
Following #210, the clicking on ramps in ramp chart generated using local repos will redirect user to non-existence link. This is because local repository lacks organization info.

Let's fix this by making the links redirect to the correct remote repository if the remote url can be found in the local repository, otherwise make it non-clickable.

**Note: an alternative solution is to show the commits info locally in the dashboard instead of redirecting to GitHub repo.

### PR DESCRIPTION
**Pull request by gok99**
Implements the suggestion in https://github.com/reposense/RepoSense/pull/717#issuecomment-497537302:
- For local repositories, attempt to find remote URL with `git remote`
- If no remote is found, disable links to the remote repository for commit view and blame view

Fixes #231 

Proposed Commit Message: 
```
Local repositories cloned from a remote or with remotes added manually
do not have working links to a remote repository, even though remote
repository information can be obtained with `git remote -v`.

Let's create a GitRemote class to obtain available remote repository 
information for local repositories, and use it for generating valid
links in a report.
```

### CODE REVIEW

**Comment by chan-j-d** [comment]
Just to record and follow up on what was discussed earlier today:

I believe a way to support this on top of PR #1644 would be to use the `RepoLocation::getRemoteRepoNameAndOrg` method but it would require changing the access modifier. 

Beside that, you might need to add separate fields to `RepoLocation`, something like `remoteRepoName` and `remoteOrg`. For remote repositories, the `remoteRepoName` and `repoName` will be the same and similarly for `remoteOrg`. For local repositories, they might be different as the output directory of the cloned local repo might have been changed. 

This is mostly related to #1637 where if the `repoName` and `organization` are the same for two directories, then their output directories generated from those two fields will be the same and the final report will override each other.

**Comment by gok99** [comment]
> Just to record and follow up on what was discussed earlier today:
> 
> I believe a way to support this on top of PR https://github.com/reposense/RepoSense/pull/1644 would be to use the RepoLocation::getRemoteRepoNameAndOrg method but it would require changing the access modifier.
> 
> Beside that, you might need to add separate fields to RepoLocation, something like remoteRepoName and remoteOrg. For remote repositories, the remoteRepoName and repoName will be the same and similarly for remoteOrg. For local repositories, they might be different as the output directory of the cloned local repo might have been changed. Then you can edit the getRepoLink in api.js to use remoteRepoName and remoteOrg instead.
> 
> This is mostly related to https://github.com/reposense/RepoSense/issues/1637 where if the repoName and organization are the same for two directories, then their output directories generated from those two fields will be the same and the final report will override each other.

I am currently working around this by adding transient fields specifically for local repositories, which store what repoName and repoOrg used to and get used for generating output directories. If a remote repository exists, then repoName and repoOrg (which get serialized and used for generating links) are obtained from the remote repository from GitRemote.

File: src/main/java/reposense/git/GitRemote.java
Comment (chan-j-d) [inline src/main/java/reposense/git/GitRemote.java:14]: Not directly related to this PR but this seems like an extremely strange line for codecov to flag as not covered.

File: src/main/java/reposense/git/GitRemote.java
Comment (chan-j-d) [inline src/main/java/reposense/git/GitRemote.java:26]: Do you think some kind of logged message is necessary here? Something like 'no remote repository was found'

File: src/main/java/reposense/model/RepoLocation.java
Comment (chan-j-d) [inline src/main/java/reposense/model/RepoLocation.java:85]: I think for lines 83 and 85, the constant `origin(fetch)` should be stored in the `GitRemote` class. Either that or create some kind of static accessor method in `GitRemote` that uses this key instead.

File: src/main/java/reposense/git/GitRemote.java
Comment (chan-j-d) [inline src/main/java/reposense/git/GitRemote.java:18]: This comment is chronologically after the comment about lines 83 and 85. Are these key names meant to be stored as constants? Because the one in the javadoc does not seem to match with the one used in lines 83 and 85.

File: src/main/java/reposense/model/RepoLocation.java
Comment (chan-j-d) [inline src/main/java/reposense/model/RepoLocation.java:115]: I was thinking that instead of storing `transient` fields `localRepoName` and `localOrganization`, we instead store `repoName` and `organization` as transient fields. Then change `localRepoName` and `localOrganization` to `remoteRepoName` and `remoteOrganization` which will be the ones uploaded via the JSON file. This way, these 2 methods do not need the `if-else` and `isLocal` field since we can always just return `repoName` and `organization`.

**Review by chan-j-d — CHANGES_REQUESTED** [review]
I've tested the enhancement a bit and it seems to work well. Some minor comments about variable usage and javadocs.

File: src/main/java/reposense/model/RepoLocation.java
Comment (gok99) [reply src/main/java/reposense/model/RepoLocation.java:85]: That's a good idea! Thanks for the suggestion.

File: src/main/java/reposense/git/GitRemote.java
Comment (gok99) [reply src/main/java/reposense/git/GitRemote.java:26]: I initially did have a log message here actually. I removed it because this catch will generally only hit when the directory isn't even a git repository. It seemed a bit strange to be logging that here, since the report generation will have to fail eventually anyway.

File: src/main/java/reposense/git/GitRemote.java
Comment (gok99) [reply src/main/java/reposense/git/GitRemote.java:14]: Oh I assumed that adding the tests would have gotten rid of this. I'm actually not sure why this is flagged.

File: src/main/java/reposense/git/GitRemote.java
Comment (gok99) [reply src/main/java/reposense/git/GitRemote.java:18]: REMOTE_NAME is meant to be substituted with the name of the remote. Lines 83 and 85 check for the remote 'origin' first if it exists. It could be possible that the repository has other remotes, Am open to suggestions for making the substitution more obvious.

File: src/main/java/reposense/model/RepoLocation.java
Comment (gok99) [reply src/main/java/reposense/model/RepoLocation.java:115]: That's probably a good idea. I was trying my best not to have to change anything outside of this class, but your suggestion seems cleaner overall.

File: src/main/java/reposense/git/GitRemote.java
Comment (chan-j-d) [reply src/main/java/reposense/git/GitRemote.java:26]: Because right now, if the command is run against a purely local repository as opposed to a non-git repository, the outcomes are the same and there is no way to distinguish. I thought it might be helpful to add in the event that down the line, some relative pathing issue results in the command being run in the wrong directory altogether. Or a fully remote repository being mistaken for a non-git directory.

File: src/main/java/reposense/git/GitRemote.java
Comment (chan-j-d) [reply src/main/java/reposense/git/GitRemote.java:18]: oh I understand it now. I think I understand the PR a lot better now. Then I believe there might be a second point of failure for repos where the user renamed the `origin` reference though I guess that is quite a fringe case and the behaviour of reposense would still be acceptable.

File: src/main/java/reposense/model/RepoLocation.java
Comment (chan-j-d) [reply src/main/java/reposense/model/RepoLocation.java:115]: Oops yea. I believe you might need to change the key references in `api.js` as well.

File: src/main/java/reposense/git/GitRemote.java
Comment (gok99) [reply src/main/java/reposense/git/GitRemote.java:31]: Yes, `git remote -v` will still succeed in a purely local git repository. The closest log message to the truth that i can think of is something like: `Unable to run git remote command in directory: %s`. Do you have any suggestions for improving this?

File: src/main/java/reposense/model/RepoLocation.java
Comment (gok99) [reply src/main/java/reposense/model/RepoLocation.java]: What do you think of the current solution?

File: src/main/java/reposense/git/GitRemote.java
Comment (gok99) [reply src/main/java/reposense/git/GitRemote.java]: Yes, if `origin` doesn't exist, we currently get whatever is available.

File: src/main/java/reposense/git/GitRemote.java
Comment (chan-j-d) [reply src/main/java/reposense/git/GitRemote.java:31]: I think that is good enough. I think this should be mostly for defensive programming since I don't expect that line to ever be reached given proper execution. But we never know 🤷

File: src/main/java/reposense/model/RepoLocation.java
Comment (chan-j-d) [reply src/main/java/reposense/model/RepoLocation.java]: I think it looks good! I'll approve the PR once you've added the logger warning.

**Review by chan-j-d — APPROVED** [review]
LGTM!

File: src/main/java/reposense/git/GitRemote.java
Comment (dcshzj) [inline src/main/java/reposense/git/GitRemote.java:67]: This part feels a little hard to read. May I suggest first checking if `remoteMap.size() == 0`, then return `Optional.empty()`? After that, we can continue with the `remoteLocation` part with the `remoteMap.containsKey(DEFAULT_FETCH_REMOTE)` check.

Something along these lines:
```java
if (remoteMap.size() == 0) {
    return Optional.empty();
}

String remoteLocation = remoteMap.containsKey(DEFAULT_FETCH_REMOTE)
...
```

File: src/main/java/reposense/model/RepoLocation.java
Comment (dcshzj) [inline src/main/java/reposense/model/RepoLocation.java:76]: Nothing to do with this line in particular, but are you also able to update the Javadoc for this `getLocalRepoNameAndOrg()` function? It is still referring to the array returned as length 2. Same for the `getRemoteRepoNameAndOrg()` function.

File: src/test/java/reposense/git/GitRemoteTest.java
Comment (dcshzj) [inline src/test/java/reposense/git/GitRemoteTest.java:13]: The name of the function should follow something like the format `featureUnderTest_testScenario_expectedBehavior()`. I think if possible, it would be nice to test `getRemotes()` and `getAvailableRemoteLocation()` separately.

**Review by dcshzj — CHANGES_REQUESTED** [review]

**Comment by gok99** [comment]
@dcshzj Updated with your suggestions changes

File: src/test/java/reposense/git/GitRemoteTest.java
Comment (dcshzj) [inline src/test/java/reposense/git/GitRemoteTest.java:65]:
```suggestion
```

Extra line here.

**Review by dcshzj — CHANGES_REQUESTED** [review]

**Review by dcshzj — APPROVED** [review]
LGTM!

**Comment by dcshzj** [comment]
Putting the merge on hold first, as there seems to be some issues with the building of the preview dashboards. @yhtMinceraft1010X Is there a chance that it is due to the Codecov action?

**Comment by yhtMinceraft1010X** [comment]
This is most likely the issue causing our surge deployments to fail: https://github.com/dawidd6/action-download-artifact/issues/147

**Comment by dcshzj** [comment]
> This is most likely the issue causing our surge deployments to fail: [dawidd6/action-download-artifact#147](https://github.com/dawidd6/action-download-artifact/issues/147)

I see, thanks for investigating! I will probably hold the merge train until tomorrow, hopefully the GitHub API issue resolves by then.

**Comment by yhtMinceraft1010X** [comment]
An update on the issue: https://github.com/dawidd6/action-download-artifact/issues/147#issuecomment-1096685055

**Pull request merged**
